File: examples/language-samples/sample.fs
// Real F#: discriminated unions, pattern matching, computation expressions.
module Billing.Invoice

open System
open System.Collections.Generic

type Status =
    | Draft
    | Sent
    | Paid

type Invoice =
    { Id: int
      Customer: string
      Amount: decimal
      Status: Status }

let describe invoice =
    match invoice.Amount with
    | a when a > 10000m -> "large"
    | a when a > 0m -> "standard"
    | _ -> "empty"

let total invoices =
    invoices |> List.sumBy (fun i -> i.Amount)

let overdue invoices =
    invoices
    |> List.filter (fun i -> i.Status = Sent)
    |> List.sortByDescending (fun i -> i.Amount)

let tryFind (cache: Dictionary<int, Invoice>) id =
    match cache.TryGetValue id with
    | true, invoice -> Some invoice
    | false, _ -> None

[<EntryPoint>]
let main argv =
    let sample = { Id = 1; Customer = "Ada"; Amount = 42m; Status = Sent }
    printfn "%s" (describe sample)
    0

File: examples/language-samples/sample.ml
(* Real OCaml: modules, variants, pattern matching, records. *)

type status = Draft | Sent | Paid

type invoice = {
  id : int;
  customer : string;
  amount : float;
  status : status;
}

module Invoice = struct
  let describe inv =
    match inv.amount with
    | a when a > 10000.0 -> "large"
    | a when a > 0.0 -> "standard"
    | _ -> "empty"

  let total invoices =
    List.fold_left (fun acc inv -> acc +. inv.amount) 0.0 invoices

  let overdue invoices =
    invoices
    |> List.filter (fun inv -> inv.status = Sent)
    |> List.sort (fun a b -> compare b.amount a.amount)

  let rec find id = function
    | [] -> None
    | inv :: rest -> if inv.id = id then Some inv else find id rest
end

let () =
  let sample = { id = 1; customer = "Ada"; amount = 42.0; status = Sent } in
  Printf.printf "%s\n" (Invoice.describe sample)
